[tools] use Java class for all java-searching behavior

andrewkolos opened the pull request:
Fixes #124252, finishing work on the umbrella tracking issue, #126126.

Essentially, after this PR, no (non-test) code should be be referencing/invoking the java home or binary paths.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

andrewkolos committed:
use Java class for all java-searching behavior

andrewkolos committed:
use Java class for all java-searching behavior

andrewkolos committed:
fix test

andrewkolos committed:
lint

andrewkolos committed:
lint

andrewkolos committed:
lint

andrewkolos committed:
lint

andrewkolos committed:
lint

andrewkolos committed:
lint

reidbaker commented on packages/flutter_tools/lib/src/android/android_workflow.dart:
```diff
@@ -75,74 +75,51 @@ class AndroidWorkflow implements Workflow {
         _logger.printTrace(error.toString());
       }
-      final Version? javaVersion = Version.parse(_extractJavaVersion(javaVersionText));
+      final Version? javaVersion = Version.parse(javaVersionNumber);
```
This can be outside the scope of the pr but I think `_java` should provide a `Version` object so that this type of conversion does not happen in multiple places slightly differently.

reidbaker commented on packages/flutter_tools/lib/src/flutter_cache.dart:
```diff
@@ -387,12 +386,17 @@ class AndroidGenSnapshotArtifacts extends EngineCachedArtifact {
 /// A cached artifact containing the Maven dependencies used to build Android projects.
 ///
 /// This is a no-op if the android SDK is not available.
```
Is this comment still true?

andrewkolos replied:
The comment appears to be true given https://github.com/flutter/flutter/blob/e8f5a51f79fc5dd2f8de38b4d9f504d09dca8197/packages/flutter_tools/lib/src/flutter_cache.dart#L411-L413, though I am not sure if it _should_. I've created https://github.com/flutter/flutter/issues/127848 to track as a follow-up and added a TODO comment linking to it

andrewkolos committed:
use Version instead of new JavaVersion

andrewkolos committed:
add todo for null check in AndroidMavenArtifacts

andrewkolos committed:
fix

andrewkolos committed:
lints

engine-flutter-autoroll committed:
Roll Flutter Engine from ee53c0499732 to 8e67cc3d6ea4 (2 revisions) (#127975)

https://github.com/flutter/engine/compare/ee53c0499732...8e67cc3d6ea4

2023-05-31 skia-flutter-autoroll@skia.org Roll Skia from f475d4f5e080 to cb883f64681b (6 revisions) (flutter/engine#42452)
2023-05-31 kjlubick@users.noreply.github.com Fix bugprone-unchecked-optional-access errors in image_generator_apng (flutter/engine#42450)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jonahwilliams@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

engine-flutter-autoroll committed:
Roll Flutter Engine from 8e67cc3d6ea4 to b9860dcc8ddd (1 revision) (#127979)

https://github.com/flutter/engine/compare/8e67cc3d6ea4...b9860dcc8ddd

2023-05-31 kjlubick@users.noreply.github.com Replace use of Skia's private GrRectanizer with a copy of the equivalent code (flutter/engine#42430)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jonahwilliams@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

andrewkolos committed:
Merge remote-tracking branch 'upstream/master' into java-finding

andrewkolos committed:
nonchange to unstuck frob
